// sanity/lib/desk.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
var structure = require('./_chunks/structure-Vsl1B9Jz.js');
require('react/jsx-runtime');
require('react');
require('@sanity/ui');
require('sanity');
require('./_chunks/Tooltip-HN_as_Fj.js');
require('@sanity/icons');
require('sanity/router');
require('@sanity/types');
require('lodash/omit.js');
require('@sanity/util/paths');
require('lodash/isNumber.js');
require('lodash/isString.js');
require('react-rx');
const ComponentBuilder = structure.ComponentBuilder;
const ComponentViewBuilder = structure.ComponentViewBuilder;
const ConfirmDeleteDialog = structure.ConfirmDeleteDialogContainer;
const DEFAULT_INTENT_HANDLER = structure.DEFAULT_INTENT_HANDLER;
const DeskToolProvider = structure.StructureToolProvider;
const DocumentBuilder = structure.DocumentBuilder;
const DocumentInspectorHeader = structure.DocumentInspectorHeader;
const DocumentListBuilder = structure.DocumentListBuilder;
const DocumentListItemBuilder = structure.DocumentListItemBuilder;
const DocumentListPane = structure.DocumentListPane;
const DocumentPane = structure.DocumentPane;
const DocumentPaneProvider = structure.DocumentPaneProvider;
const DocumentTypeListBuilder = structure.DocumentTypeListBuilder;
const FormViewBuilder = structure.FormViewBuilder;
const GenericListBuilder = structure.GenericListBuilder;
const GenericViewBuilder = structure.GenericViewBuilder;
const HELP_URL = structure.HELP_URL;
const InitialValueTemplateItemBuilder = structure.InitialValueTemplateItemBuilder;
const ListBuilder = structure.ListBuilder;
const ListItemBuilder = structure.ListItemBuilder;
const MenuItemBuilder = structure.MenuItemBuilder;
const MenuItemGroupBuilder = structure.MenuItemGroupBuilder;
const PaneLayout = structure.PaneLayout;
const PaneRouterContext = structure.PaneRouterContext;
const SerializeError = structure.SerializeError;
const component = structure.component;
const createStructureBuilder = structure.createStructureBuilder;
const defaultInitialValueTemplateItems = structure.defaultInitialValueTemplateItems;
const defaultIntentChecker = structure.defaultIntentChecker;
const deskTool = structure.structureTool;
const documentFromEditor = structure.documentFromEditor;
const documentFromEditorWithInitialValue = structure.documentFromEditorWithInitialValue;
const form = structure.form;
const getOrderingMenuItem = structure.getOrderingMenuItem;
const getOrderingMenuItemsForSchemaType = structure.getOrderingMenuItemsForSchemaType;
const getTypeNamesFromFilter = structure.getTypeNamesFromFilter;
const isDocumentListItem = structure.isDocumentListItem;
const maybeSerializeInitialValueTemplateItem = structure.maybeSerializeInitialValueTemplateItem;
const maybeSerializeMenuItem = structure.maybeSerializeMenuItem;
const maybeSerializeMenuItemGroup = structure.maybeSerializeMenuItemGroup;
const maybeSerializeView = structure.maybeSerializeView;
const menuItemsFromInitialValueTemplateItems = structure.menuItemsFromInitialValueTemplateItems;
const shallowIntentChecker = structure.shallowIntentChecker;
const structureLocaleNamespace = structure.structureLocaleNamespace;
const useDeskTool = structure.useStructureTool;
const useDocumentPane = structure.useDocumentPane;
const useDocumentTitle = structure.useDocumentTitle;
const usePaneRouter = structure.usePaneRouter;
exports.ComponentBuilder = ComponentBuilder;
exports.ComponentViewBuilder = ComponentViewBuilder;
exports.ConfirmDeleteDialog = ConfirmDeleteDialog;
exports.DEFAULT_INTENT_HANDLER = DEFAULT_INTENT_HANDLER;
exports.DeskToolProvider = DeskToolProvider;
exports.DocumentBuilder = DocumentBuilder;
exports.DocumentInspectorHeader = DocumentInspectorHeader;
exports.DocumentListBuilder = DocumentListBuilder;
exports.DocumentListItemBuilder = DocumentListItemBuilder;
exports.DocumentListPane = DocumentListPane;
exports.DocumentPane = DocumentPane;
exports.DocumentPaneProvider = DocumentPaneProvider;
exports.DocumentTypeListBuilder = DocumentTypeListBuilder;
exports.FormViewBuilder = FormViewBuilder;
exports.GenericListBuilder = GenericListBuilder;
exports.GenericViewBuilder = GenericViewBuilder;
exports.HELP_URL = HELP_URL;
exports.InitialValueTemplateItemBuilder = InitialValueTemplateItemBuilder;
exports.ListBuilder = ListBuilder;
exports.ListItemBuilder = ListItemBuilder;
exports.MenuItemBuilder = MenuItemBuilder;
exports.MenuItemGroupBuilder = MenuItemGroupBuilder;
exports.PaneLayout = PaneLayout;
exports.PaneRouterContext = PaneRouterContext;
exports.SerializeError = SerializeError;
exports.component = component;
exports.createStructureBuilder = createStructureBuilder;
exports.defaultInitialValueTemplateItems = defaultInitialValueTemplateItems;
exports.defaultIntentChecker = defaultIntentChecker;
exports.deskTool = deskTool;
exports.documentFromEditor = documentFromEditor;
exports.documentFromEditorWithInitialValue = documentFromEditorWithInitialValue;
exports.form = form;
exports.getOrderingMenuItem = getOrderingMenuItem;
exports.getOrderingMenuItemsForSchemaType = getOrderingMenuItemsForSchemaType;
exports.getTypeNamesFromFilter = getTypeNamesFromFilter;
exports.isDocumentListItem = isDocumentListItem;
exports.maybeSerializeInitialValueTemplateItem = maybeSerializeInitialValueTemplateItem;
exports.maybeSerializeMenuItem = maybeSerializeMenuItem;
exports.maybeSerializeMenuItemGroup = maybeSerializeMenuItemGroup;
exports.maybeSerializeView = maybeSerializeView;
exports.menuItemsFromInitialValueTemplateItems = menuItemsFromInitialValueTemplateItems;
exports.shallowIntentChecker = shallowIntentChecker;
exports.structureLocaleNamespace = structureLocaleNamespace;
exports.useDeskTool = useDeskTool;
exports.useDocumentPane = useDocumentPane;
exports.useDocumentTitle = useDocumentTitle;
exports.usePaneRouter = usePaneRouter;
