// @sanity/client/dist/csm.cjs
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
var resolveEditInfo = require('./_chunks/resolveEditInfo-TYjTkHKU.cjs');
const defaultUpdateFunction = changed => changed;
function applySourceDocuments(result, resultSourceMap, getCachedDocument) {
  let updateFn = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : defaultUpdateFunction;
  let perspective = arguments.length > 4 && arguments[4] !== undefined ? arguments[4] : "raw";
  if (!resultSourceMap) return result;
  if (perspective !== "published" && perspective !== "raw" && perspective !== "previewDrafts") {
    throw new Error('Unknown perspective "'.concat(perspective, '"'));
  }
  return resolveEditInfo.walkMap(JSON.parse(JSON.stringify(result)), (value, path) => {
    const resolveMappingResult = resolveEditInfo.resolveMapping(path, resultSourceMap);
    if (!resolveMappingResult) {
      return value;
    }
    const {
      mapping,
      pathSuffix
    } = resolveMappingResult;
    if (mapping.type !== "value") {
      return value;
    }
    if (mapping.source.type !== "documentValue") {
      return value;
    }
    const sourceDocument = resultSourceMap.documents[mapping.source.document];
    const sourcePath = resultSourceMap.paths[mapping.source.path];
    if (sourceDocument) {
      const parsedPath = resolveEditInfo.parseJsonPath(sourcePath + pathSuffix);
      const stringifiedPath = resolveEditInfo.toString(parsedPath);
      if (stringifiedPath === "_id") {
        return value;
      }
      let cachedDocument;
      if (perspective === "previewDrafts") {
        cachedDocument = getCachedDocument(sourceDocument._id.startsWith(resolveEditInfo.DRAFTS_PREFIX) ? sourceDocument : {
          ...sourceDocument,
          _id: "".concat(resolveEditInfo.DRAFTS_PREFIX, ".").concat(sourceDocument._id, "}")
        });
        if (!cachedDocument) {
          cachedDocument = getCachedDocument(sourceDocument._id.startsWith(resolveEditInfo.DRAFTS_PREFIX) ? {
            ...sourceDocument,
            _id: resolveEditInfo.getPublishedId(sourceDocument._id)
          } : sourceDocument);
        }
        if (cachedDocument) {
          cachedDocument = {
            ...cachedDocument,
            _id: resolveEditInfo.getPublishedId(sourceDocument._id),
            _originalId: sourceDocument._id
          };
        }
      } else {
        cachedDocument = getCachedDocument(sourceDocument);
      }
      if (!cachedDocument) {
        return value;
      }
      const changedValue = cachedDocument ? resolveEditInfo.get(cachedDocument, stringifiedPath, value) : value;
      return value === changedValue ? value : updateFn(changedValue, {
        cachedDocument,
        previousValue: value,
        sourceDocument,
        sourcePath: parsedPath
      });
    }
    return value;
  });
}
function resolvedKeyedSourcePath(options) {
  const {
    keyedResultPath,
    pathSuffix,
    sourceBasePath
  } = options;
  const inferredResultPath = pathSuffix === void 0 ? [] : resolveEditInfo.parseJsonPath(pathSuffix);
  const inferredPath = keyedResultPath.slice(keyedResultPath.length - inferredResultPath.length);
  const inferredPathSuffix = inferredPath.length ? resolveEditInfo.jsonPath(inferredPath).slice(1) : "";
  return resolveEditInfo.parseJsonPath(sourceBasePath + inferredPathSuffix);
}
function resolveEditUrl(options) {
  const {
    resultSourceMap,
    studioUrl
  } = options;
  const resultPath = resolveEditInfo.studioPathToJsonPath(options.resultPath);
  const editInfo = resolveEditInfo.resolveEditInfo({
    resultPath,
    resultSourceMap,
    studioUrl
  });
  if (!editInfo) {
    return void 0;
  }
  return resolveEditInfo.createEditUrl(editInfo);
}
exports.createEditUrl = resolveEditInfo.createEditUrl;
exports.getPublishedId = resolveEditInfo.getPublishedId;
exports.jsonPath = resolveEditInfo.jsonPath;
exports.jsonPathToStudioPath = resolveEditInfo.jsonPathToStudioPath;
exports.parseJsonPath = resolveEditInfo.parseJsonPath;
exports.resolveEditInfo = resolveEditInfo.resolveEditInfo;
exports.resolveMapping = resolveEditInfo.resolveMapping;
exports.studioPath = resolveEditInfo.studioPath;
exports.studioPathToJsonPath = resolveEditInfo.studioPathToJsonPath;
exports.walkMap = resolveEditInfo.walkMap;
exports.applySourceDocuments = applySourceDocuments;
exports.resolveEditUrl = resolveEditUrl;
exports.resolvedKeyedSourcePath = resolvedKeyedSourcePath;
