// @sanity/client/dist/csm.js
import { walkMap, resolveMapping, parseJsonPath, toString, DRAFTS_PREFIX, getPublishedId, get, jsonPath, studioPathToJsonPath, resolveEditInfo, createEditUrl } from './_chunks/resolveEditInfo-FaeuCV4M.js';
export { jsonPathToStudioPath, studioPath } from './_chunks/resolveEditInfo-FaeuCV4M.js';
const defaultUpdateFunction = changed => changed;
function applySourceDocuments(result, resultSourceMap, getCachedDocument) {
  let updateFn = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : defaultUpdateFunction;
  let perspective = arguments.length > 4 && arguments[4] !== undefined ? arguments[4] : "raw";
  if (!resultSourceMap) return result;
  if (perspective !== "published" && perspective !== "raw" && perspective !== "previewDrafts") {
    throw new Error('Unknown perspective "'.concat(perspective, '"'));
  }
  return walkMap(JSON.parse(JSON.stringify(result)), (value, path) => {
    const resolveMappingResult = resolveMapping(path, resultSourceMap);
    if (!resolveMappingResult) {
      return value;
    }
    const {
      mapping,
      pathSuffix
    } = resolveMappingResult;
    if (mapping.type !== "value") {
      return value;
    }
    if (mapping.source.type !== "documentValue") {
      return value;
    }
    const sourceDocument = resultSourceMap.documents[mapping.source.document];
    const sourcePath = resultSourceMap.paths[mapping.source.path];
    if (sourceDocument) {
      const parsedPath = parseJsonPath(sourcePath + pathSuffix);
      const stringifiedPath = toString(parsedPath);
      if (stringifiedPath === "_id") {
        return value;
      }
      let cachedDocument;
      if (perspective === "previewDrafts") {
        cachedDocument = getCachedDocument(sourceDocument._id.startsWith(DRAFTS_PREFIX) ? sourceDocument : {
          ...sourceDocument,
          _id: "".concat(DRAFTS_PREFIX, ".").concat(sourceDocument._id, "}")
        });
        if (!cachedDocument) {
          cachedDocument = getCachedDocument(sourceDocument._id.startsWith(DRAFTS_PREFIX) ? {
            ...sourceDocument,
            _id: getPublishedId(sourceDocument._id)
          } : sourceDocument);
        }
        if (cachedDocument) {
          cachedDocument = {
            ...cachedDocument,
            _id: getPublishedId(sourceDocument._id),
            _originalId: sourceDocument._id
          };
        }
      } else {
        cachedDocument = getCachedDocument(sourceDocument);
      }
      if (!cachedDocument) {
        return value;
      }
      const changedValue = cachedDocument ? get(cachedDocument, stringifiedPath, value) : value;
      return value === changedValue ? value : updateFn(changedValue, {
        cachedDocument,
        previousValue: value,
        sourceDocument,
        sourcePath: parsedPath
      });
    }
    return value;
  });
}
function resolvedKeyedSourcePath(options) {
  const {
    keyedResultPath,
    pathSuffix,
    sourceBasePath
  } = options;
  const inferredResultPath = pathSuffix === void 0 ? [] : parseJsonPath(pathSuffix);
  const inferredPath = keyedResultPath.slice(keyedResultPath.length - inferredResultPath.length);
  const inferredPathSuffix = inferredPath.length ? jsonPath(inferredPath).slice(1) : "";
  return parseJsonPath(sourceBasePath + inferredPathSuffix);
}
function resolveEditUrl(options) {
  const {
    resultSourceMap,
    studioUrl
  } = options;
  const resultPath = studioPathToJsonPath(options.resultPath);
  const editInfo = resolveEditInfo({
    resultPath,
    resultSourceMap,
    studioUrl
  });
  if (!editInfo) {
    return void 0;
  }
  return createEditUrl(editInfo);
}
export { applySourceDocuments, createEditUrl, getPublishedId, jsonPath, parseJsonPath, resolveEditInfo, resolveEditUrl, resolveMapping, resolvedKeyedSourcePath, studioPathToJsonPath, walkMap };
