// @sanity/client/dist/index.cjs
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
var nodeMiddleware = require('./_chunks/nodeMiddleware-qKdVvFDs.cjs');
var getIt = require('get-it');
function defineDeprecatedCreateClient(createClient) {
  return function deprecatedCreateClient(config) {
    nodeMiddleware.printNoDefaultExport();
    return createClient(config);
  };
}
const exp = nodeMiddleware.defineCreateClientExports(nodeMiddleware.middleware, nodeMiddleware.SanityClient);
const requester = exp.requester;
const createClient = exp.createClient;
const deprecatedCreateClient = defineDeprecatedCreateClient(createClient);
exports.BasePatch = nodeMiddleware.BasePatch;
exports.BaseTransaction = nodeMiddleware.BaseTransaction;
exports.ClientError = nodeMiddleware.ClientError;
exports.ObservablePatch = nodeMiddleware.ObservablePatch;
exports.ObservableSanityClient = nodeMiddleware.ObservableSanityClient;
exports.ObservableTransaction = nodeMiddleware.ObservableTransaction;
exports.Patch = nodeMiddleware.Patch;
exports.SanityClient = nodeMiddleware.SanityClient;
exports.ServerError = nodeMiddleware.ServerError;
exports.Transaction = nodeMiddleware.Transaction;
Object.defineProperty(exports, "unstable__adapter", {
  enumerable: true,
  get: function () {
    return getIt.adapter;
  }
});
Object.defineProperty(exports, "unstable__environment", {
  enumerable: true,
  get: function () {
    return getIt.environment;
  }
});
exports.createClient = createClient;
exports.default = deprecatedCreateClient;
exports.requester = requester;
